app/presenters/tree_builder.rb

```diff
@@ -113,6 +120,13 @@ def build_tree
     @locals_for_render = set_locals_for_render
   end
 
+  def opened_nodes
+    # If the open_nodes is not set, Array(nil) will always return a different object, therefore,
+    # for stateless trees it has a performance drawback of creating a new array after each time
+    # this method is called. By memoizing the first result, only a single array gets allocated.
+    @opened_nodes ||= Array(@tree_state.x_tree(@name)[:open_nodes])
```

Comment:
This will break any code which updates `:open_nodes` in tree state from the outside, right?

```
app/controllers/miq_ae_class_controller.rb:261:           x_tree[:open_nodes].include?(x_build_node_id(parents.last))
app/controllers/miq_ae_class_controller.rb:264:        if x_tree[:open_nodes].include?(p_node)
app/controllers/miq_ae_class_controller.rb:267:          x_tree[:open_nodes].push(p_node)
app/controllers/mixins/sandbox.rb:24:      :open_nodes => []
app/controllers/catalog_controller.rb:733:    x_tree[:open_nodes].push("xx-#{ot_type}") unless x_tree[:open_nodes].include?("xx-#{ot_type}")
app/controllers/catalog_controller.rb:735:    x_tree[:open_nodes].push(x_node)
app/controllers/catalog_controller.rb:1766:           x_tree[:open_nodes].include?(parents.last[:id])
app/controllers/catalog_controller.rb:1770:        unless x_tree[:open_nodes].include?(p_node)
app/controllers/catalog_controller.rb:1771:          x_tree[:open_nodes].push(p_node)
app/controllers/report_controller.rb:619:           x_tree[:open_nodes].include?(x_build_node_id(parents.last))
app/controllers/report_controller.rb:624:        x_tree[:open_nodes].push(p_node) unless x_tree[:open_nodes].include?(p_node)
app/controllers/ops_controller.rb:586:    unless parents.empty? || x_tree[:open_nodes].include?(x_build_node_id(parents.last))
app/controllers/ops_controller.rb:589:        unless x_tree[:open_nodes].include?(p_node)
app/controllers/ops_controller.rb:590:          x_tree[:open_nodes].push(p_node)
```

Reply:
Yeah, this would be true if we'd be adjusting the `x_tree[:open_nodes]` during the `TreeBuilder` is "alive". As far as I can see, we don't really do that as the `TreeBuilder` instantly builds the tree and its instance is never reused again.

Reply:
Aaah, makes sense, thanks :) :+1: 

Well, the instance is never reused but the tree state is saved in session...
But as long as calls to `open_node` still end up in that state (and they do seem to), we're fine  :).

Reply:
Exactly :+1: